### PR TITLE
Feature/wordpress cphs

### PR DESCRIPTION
**Pull request by bogdanguranda**
Implementation:
- kafka listener now ingests messages and sends 2 messages, one to content, and another to complementarycontent collection
- /map endpoint now returns array of the 2 messages (content, and complementarycontent collections)
- complementarycontent messages contains all promotional fields
- checks for internal CPHs and external CPHs and writes messages to kafka accordingly
- refactored code to separate distinct parts like, model, validation, errors, mapping logic (also using golang composition for shared bits in the model)

### CODE REVIEW

File: mapper/mapper.go [hunk 1]
```diff
@@ -1,203 +1,116 @@
...
 }
 
-type mapper struct {
+type defaultMapper struct {
 	messageConsumer consumer.MessageConsumer
 	messageProducer producer.MessageProducer
+	cphValidator    CPHValidator
 }
 
 // New returns a new Mapper instance
```
Comment (kovacshuni) [inline mapper/mapper.go]: Remove the comment please.

File: mapper/validator.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+func NewDefaultCPHValidator() *defaultCPHValidator {
+	return &defaultCPHValidator{}
+}
+
+func (dcv *defaultCPHValidator) Validate(mcp *model.MethodeContentPlaceholder) error {
+	return dcv.validateHeadline(mcp.Body.LeadHeadline)
+}
+
+func (dcv *defaultCPHValidator) validateHeadline(headline model.LeadHeadline) error {
```
Comment (MihaiSorinMoisa) [inline mapper/validator.go]: Is it necessary for this function to have a receiver?

File: mapper/validator.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package mapper
+
+import (
```
Comment (MihaiSorinMoisa) [inline mapper/validator.go]: Please organize the imports the following way https://github.com/golang/go/wiki/CodeReviewComments#imports.

File: mapper/mapper_test.go [hunk 4]
```diff
@@ -14,25 +12,33 @@ import (
 	"github.com/Financial-Times/message-queue-gonsumer/consumer"
```
Comment (MihaiSorinMoisa) [inline mapper/mapper_test.go]: Please organize the imports the following way https://github.com/golang/go/wiki/CodeReviewComments#imports.

File: mapper/mapper_test.go [hunk 5]
```diff
@@ -1,10 +1,8 @@
 package mapper
 
 import (
```
Comment (MihaiSorinMoisa) [inline mapper/mapper_test.go]: Please organize the imports the following way https://github.com/golang/go/wiki/CodeReviewComments#imports.

File: mapper/mapper.go [hunk 6]
```diff
@@ -1,203 +1,116 @@
...
-}
 
-func buildAlternativeImages(fileRef string) *AlternativeImages {
-	if fileRef == "" {
-		return nil
-	}
-	imageUUID := extractImageUUID(fileRef)
-	return &AlternativeImages{PromotionalImage: imageUUID}
-}
+	if mcp.IsInternalCPH() {
```
Comment (MihaiSorinMoisa) [inline mapper/mapper.go]: This flow should be tested as well.

File: mapper/mapper.go [hunk 7]
```diff
@@ -1,203 +1,116 @@
...
 
-func buildAlternativeStandfirsts(promoStandfirst string) *AlternativeStandfirsts {
-	promoStandfirst = strings.TrimSpace(promoStandfirst)
-	if promoStandfirst == "" {
-		return nil
-	}
-	return &AlternativeStandfirsts{PromotionalStandfirst: promoStandfirst}
-}
+		uppContent, err := model.NewUppContentPlaceholder(mcp)
+		if err != nil {
```
Comment (MihaiSorinMoisa) [inline mapper/mapper.go]: This flow should be tested as well.

File: mapper/mapper.go [hunk 8]
```diff
@@ -1,203 +1,116 @@
...
-	upPlaceholder, err := m.MapContentPlaceholder(methodePlaceholder)
-	if err != nil {
-		return producer.Message{}, "", err
+		log.WithField("transaction_id", tid).WithField("uuid", placeholderUUID).WithError(err).Warn("Error sending transformed content message to queue")
+		return
 	}
 
-	pubEventMsg, err := upPlaceholder.toPublicationEventMessage()
+	err = m.messageProducer.SendMessage(complementaryUUID, *complementaryMsg)
 	if err != nil {
```
Comment (MihaiSorinMoisa) [inline mapper/mapper.go]: This flow should be tested as well.

File: mapper/mapper.go [hunk 9]
```diff
@@ -1,203 +1,116 @@
...
-	return &mapper{}
+func NewDefaultMapper() *defaultMapper {
+	return &defaultMapper{cphValidator: NewDefaultCPHValidator()}
 }
 
-func (m *mapper) HandlePlaceholderMessages(msg consumer.Message) {
+func (m *defaultMapper) HandlePlaceholderMessages(msg consumer.Message) {
 	tid := msg.Headers["X-Request-Id"]
-	if msg.Headers["Origin-System-Id"] != methodeSystemID {
+	if msg.Headers["Origin-System-Id"] != model.MethodeSystemID {
```
Comment (MihaiSorinMoisa) [inline mapper/mapper.go]: This flow should be tested as well.

File: mapper/mapper.go [hunk 10]
```diff
@@ -1,203 +1,116 @@
...
 		log.WithField("transaction_id", tid).WithField("Origin-System-Id", msg.Headers["Origin-System-Id"]).Info("Ignoring message with different Origin-System-Id")
 		return
 	}
-	placeholderMsg, placeholderUUID, mappingErr := m.mapMessage(msg)
+
+	placeholderMsg, placeholderUUID, complementaryMsg, complementaryUUID, mappingErr := m.mapMessage(msg)
 	if mappingErr != nil {
```
Comment (MihaiSorinMoisa) [inline mapper/mapper.go]: This flow should be tested as well.

File: mapper/mapper.go [hunk 11]
```diff
@@ -1,203 +1,116 @@
...
-	if mpc.attributes.IsDeleted {
-		return UpContentPlaceholder{
-			UUID:             mpc.UUID,
-			LastModified:     mpc.lastModified,
-			PublishReference: mpc.transactionID,
-			IsMarkedDeleted:  true,
-		}, nil
-	}
-	err := validateHeadline(mpc.body.LeadHeadline)
+func (m *defaultMapper) mapMessage(msg consumer.Message) (*producer.Message, string, *producer.Message, string, *utility.MappingError) {
```
Comment (MihaiSorinMoisa) [inline mapper/mapper.go]: None of the `if` branches are being covered be tests.

File: model/methodeContentPlaceholder.go [hunk 12]
```diff
@@ -0,0 +1,110 @@
+package model
+
+import (
```
Comment (MihaiSorinMoisa) [inline model/methodeContentPlaceholder.go]: Reorganize the imports

File: model/methodeContentPlaceholder.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+	Text string `xml:",chardata"`
+	URL  string `xml:"href,attr"`
+}
+
+// LeadImage represents the image attribute of a methode content placeholder
+type LeadImage struct {
+	FileRef string `xml:"fileref,attr"`
+}
+
+func (mcp MethodeContentPlaceholder) BuildAttributes(attributesXML string) (Attributes, error) {
```
Comment (MihaiSorinMoisa) [inline model/methodeContentPlaceholder.go]: Is it necessary for this function to have a receiver and be exported outside of the package?

File: model/methodeContentPlaceholder.go [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+	var attrs Attributes
+	if err := xml.Unmarshal([]byte(attributesXML), &attrs); err != nil {
+		return Attributes{}, err
+	}
+	return attrs, nil
+}
+
+func (mcp MethodeContentPlaceholder) BuildMethodeBody(methodeBodyXMLBase64 string) (MethodeBody, error) {
```
Comment (MihaiSorinMoisa) [inline model/methodeContentPlaceholder.go]: Is it necessary for this function to have a receiver and be exported outside of the package?

File: model/uppContentPlaceholder.go [hunk 15]
```diff
@@ -0,0 +1,115 @@
+package model
+
+import (
```
Comment (MihaiSorinMoisa) [inline model/uppContentPlaceholder.go]: Reorganize the imports.

File: model/uppCoreContent.go [hunk 16]
```diff
@@ -0,0 +1,61 @@
+package model
+
+import (
```
Comment (MihaiSorinMoisa) [inline model/uppCoreContent.go]: Reorganize the imports.

File: model/methodeContentPlaceholder.go [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+		return MethodeBody{}, err
+	}
+	return body, nil
+}
+
+func (mcp MethodeContentPlaceholder) IsInternalCPH() bool {
+	return mcp.Attributes.LinkedArticleUUID != ""
+}
+
+func NewMethodeContentPlaceholder(messageBody []byte, transactionID string, lastModified string) (*MethodeContentPlaceholder, *utility.MappingError) {
```
Comment (MihaiSorinMoisa) [inline model/methodeContentPlaceholder.go]: This function should be covered by tests.

File: resources/content_transform.go [hunk 18]
```diff
@@ -9,6 +9,10 @@ import (
 	log "github.com/Sirupsen/logrus"
```
Comment (MihaiSorinMoisa) [inline resources/content_transform.go]: Reorganize imports.

File: resources/content_transform.go [hunk 19]
```diff
@@ -20,6 +24,9 @@ type msg struct {
 	Message string `json:"message"`
 }
 
+type mapResponse struct {
+}
+
 // NewContentTransformHandler returns a new instance of a MapEndpointHandler
```
Comment (MihaiSorinMoisa) [inline resources/content_transform.go]: Remove this comment

File: resources/content_transform.go [hunk 20]
```diff
@@ -32,29 +39,47 @@ func (h *MapEndpointHandler) ServeMapEndpoint(w http.ResponseWriter, r *http.Req
...
 	if err != nil {
-		writeError(w, err, transactionID, uuid, r.RequestURI)
+		writeError(w, err, transactionID, "", r.RequestURI)
 		return
 	}
-	upPlaceholder, err := h.mapper.MapContentPlaceholder(methodePlaceholder)
+	uuid := methodePlaceholder.UUID
+
+	uppPlaceholder, uppComplementaryContent, err := h.mapper.MapContentPlaceholder(methodePlaceholder)
 	if err != nil {
```
Comment (MihaiSorinMoisa) [inline resources/content_transform.go]: This flow should be covered by tests

File: resources/content_transform.go [hunk 21]
```diff
@@ -32,29 +39,47 @@ func (h *MapEndpointHandler) ServeMapEndpoint(w http.ResponseWriter, r *http.Req
...
 	}
 
 	w.Header().Add("Content-Type", "application/json")
 	encoder := json.NewEncoder(w)
-	encoder.Encode(upPlaceholder)
+	encoder.Encode(h.createPublicationEventsMessages(uppPlaceholder, uppComplementaryContent))
 	log.WithField("transaction_id", transactionID).WithField("uuid", uuid).WithField("request_uri", r.RequestURI).Info("Transformation successful")
 }
 
+func (h *MapEndpointHandler) NewMethodeContentPlaceholderFromHTTPRequest(r *http.Request) (*model.MethodeContentPlaceholder, *utility.MappingError) {
```
Comment (MihaiSorinMoisa) [inline resources/content_transform.go]: I think you should remove the receiver and move this function to the `methodeContentPlaceholder` file.

File: resources/content_transform.go [hunk 22]
```diff
@@ -32,29 +39,47 @@ func (h *MapEndpointHandler) ServeMapEndpoint(w http.ResponseWriter, r *http.Req
...
+	transactionID := tid.GetTransactionIDFromRequest(r)
+	lastModified := time.Now().Format(model.UPPDateFormat)
+	messageBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return nil, utility.NewMappingError().WithMessage(err.Error())
+	}
+	return model.NewMethodeContentPlaceholder(messageBody, transactionID, lastModified)
+}
+
+func (h *MapEndpointHandler) createPublicationEventsMessages(uppPlaceholder *model.UppContentPlaceholder, uppComplementaryContent *model.UppComplementaryContent) [2]model.PublicationEvent {
```
Comment (MihaiSorinMoisa) [inline resources/content_transform.go]: You should remove the receiver from this function.

File: mapper/validator.go (hunk 2 above)
Comment (kovacshuni) [reply mapper/validator.go]: I think it's okay. It's mostly due to encapsulations, so the validation function can be mocked in places where the validator is a dependency. In `mapper.go` you could inject a mocked validator.

File: mapper/mapper.go (hunk 6 above)
Comment (kovacshuni) [reply mapper/mapper.go]: will take a look

File: mapper/mapper.go (hunk 7 above)
Comment (kovacshuni) [reply mapper/mapper.go]: will take a look

**Comment by kovacshuni** [comment]
Code and tests look ✅ 👌

**Comment by kovacshuni** [comment]
still remains Helm and `README.md`

**Comment by kovacshuni** [comment]
helm looks 🆗 , readme remains, but let's see this in action first.

**Pull request merged**